File: src/falcon_core/physics/config/geometries/gate_geometry_array_1D.py
"""An array of gates forming a 1D current channel."""

from typing import TYPE_CHECKING

from .barrier_gate_with_neighbors import BarrierGateWithNeighbors
from .dependancies import (
    BarrierGate,
    DotGate,
    Gates,
    Ohmic,
    PlungerGate,
    ReservoirGate,
)
from .left_reservoir_with_implanted_ohmic import LeftReservoirWithImplantedOhmic
from .plunger_gate_with_neighbors import PlungerGateWithNeighbors
from .right_reservoir_with_implanted_ohmic import RightReservoirWithImplantedOhmic
from .typing import (
    Ohmics,
)

if TYPE_CHECKING:
    from .typing import (
        Connection,
        Gate,
        ScreeningGates,
    )


class GateGeometryArray1D:
    """An assumption made about the geometry of the gates in the device.

    The assumed geometry is a 1D array:
                            Screening Gate
        Ohmic, Reservoir, DotGates ... DotGates, Reservoir, Ohmic
                            Screening Gate

    Located around that channel are screening gates.
    """

    left_reservoir: "LeftReservoirWithImplantedOhmic"
    right_reservoir: "RightReservoirWithImplantedOhmic"
    left_barrier: "BarrierGateWithNeighbors"
    right_barrier: "BarrierGateWithNeighbors"
    central_dot_gates: list["PlungerGateWithNeighbors | BarrierGateWithNeighbors"]
    raw_gates: "Gates"
    ohmics: "Ohmics"
    screening_gates: "ScreeningGates"
    lineararray: "list[Connection]"  # allows custom parsing

    def __init__(
        self,
        lineararray: "list[Connection]",
        screening_gates: "ScreeningGates",
    ):
        """Starts the geometry of the gates.

        Args:
            lineararray (Connections): The gates in the channel.
            screening_gates (ScreeningGates): The screening gates
            bottom_screening_gate (ScreeningGate): The bottom screening gate.
        """
        if len(screening_gates) != 2:
            msg = "Expected two screening gates."
            raise ValueError(msg)
        self.screening_gates = screening_gates
        self.lineararray = lineararray
        if len(lineararray) % 2 == 0:
            msg = "Expected an odd number of elements in the linear array."
            raise ValueError(msg)

        if (not isinstance(lineararray[0], Ohmic)) or (
            not isinstance(lineararray[-1], Ohmic)
        ):
            msg = "Expected Ohmic at the ends of the linear array."
            raise TypeError(msg)

        self.ohmics = Ohmics([lineararray[0], lineararray[-1]])

        if (not isinstance(lineararray[1], ReservoirGate)) or (
            not isinstance(lineararray[-2], ReservoirGate)
        ):
            msg = "Expected Reservoir Gates bounding the central dot gates."
            raise TypeError(msg)

        dot_gates = lineararray[2:-2]

        self.store_raw_central_gates(dot_gates=dot_gates)

        if not all([isinstance(gate, DotGate) for gate in dot_gates]):
            msg = "Expected DotGates in the middle of the linear array."
            raise TypeError(msg)

        if (not isinstance(dot_gates[0], BarrierGate)) or (
            not isinstance(dot_gates[-1], BarrierGate)
        ):
            msg = "Expected Barrier Gates bounding the exterior reservoir gates."
            raise TypeError(msg)

        if (not isinstance(dot_gates[1], PlungerGate)) or (
            not isinstance(dot_gates[-2], PlungerGate)
        ):
            msg = "Expected Barrier Gates bounding the exterior reservoir gates."
            raise TypeError(msg)

        self.left_reservoir = LeftReservoirWithImplantedOhmic(
            name=lineararray[1].name,
            ohmic=lineararray[0],
            right_neighbor=dot_gates[0],
        )
        self.right_reservoir = RightReservoirWithImplantedOhmic(
            name=lineararray[-2].name,
            ohmic=lineararray[-1],
            left_neighbor=dot_gates[-1],
        )

        self.left_barrier = BarrierGateWithNeighbors(
            name=dot_gates[0].name,
            left_neighbor=lineararray[1],
            right_neighbor=dot_gates[1],
        )

        self.right_barrier = BarrierGateWithNeighbors(
            name=dot_gates[-1].name,
            left_neighbor=dot_gates[-2],
            right_neighbor=lineararray[-2],
        )

        self.central_dot_gates = []
        for i in range(3, len(lineararray) - 3):
            selected_gate = lineararray[i]
            left_neighbor = lineararray[i - 1]
            right_neighbor = lineararray[i + 1]
            assert isinstance(selected_gate, DotGate)
            assert isinstance(left_neighbor, DotGate)
            assert isinstance(right_neighbor, DotGate)
            self.store_central_gate(
                selected_gates=[left_neighbor, selected_gate, right_neighbor],
            )

    def store_raw_central_gates(self, dot_gates: "list[Connection]") -> None:
        """Stores the raw gates in the geometry.

        Args:
            dot_gates (Connections): The dot gates.
        """
        out = []
        for gate in dot_gates:
            if isinstance(gate, DotGate):
                out.append(gate)
            else:
                msg = "Expected DotGates in the middle of the linear array."
                raise TypeError(msg)
        self.raw_gates = Gates(out)

    def store_central_gate(
        self,
        selected_gates: list[DotGate],
    ) -> None:
        """Stores a central gate in the geometry.

        Each array should contain 3 elements, the left neighbor, the selected gate, and the right neighbor.

        Args:
            selected_gates (list[DotGate, DotGate, DotGate]): The selected gates.
            selected_types (list[PlungerGate | BarrierGate]): The selected types.
        """
        left_neighbor = selected_gates[0]
        selected_gate = selected_gates[1]
        right_neighbor = selected_gates[2]
        if isinstance(selected_gate, BarrierGate):
            if not isinstance(left_neighbor, PlungerGate):
                msg = "Expected PlungerGate bounding selected BarrierGate."
                raise TypeError(msg)
            if not isinstance(right_neighbor, PlungerGate):
                msg = "Expected PlungerGate bounding selected BarrierGate."
                raise TypeError(msg)
            self.central_dot_gates.append(
                BarrierGateWithNeighbors(
                    name=selected_gate.name,
                    left_neighbor=left_neighbor,
                    right_neighbor=right_neighbor,
                )
            )
        elif isinstance(selected_gate, PlungerGate):
            if not isinstance(left_neighbor, BarrierGate):
                msg = "Expected BarrierGate bounding selected PlungerGate."
                raise TypeError(msg)
            if not isinstance(right_neighbor, BarrierGate):
                msg = "Expected BarrierGate bounding selected PlungerGate."
                raise TypeError(msg)
            self.central_dot_gates.append(
                PlungerGateWithNeighbors(
                    name=selected_gate.name,
                    left_neighbor=left_neighbor,
                    right_neighbor=right_neighbor,
                )
            )
        else:
            msg = "Expected either a PlungerGate or BarrierGate."
            raise TypeError(msg)

    def all_dot_gates(
        self,
    ) -> list["PlungerGateWithNeighbors | BarrierGateWithNeighbors"]:
        """Gets all of the dot gates in the geometry.

        Returns:
            list[PlungerGateWithNeighbors | BarrierGateWithNeighbors]: The dot gates.
        """
        return [
            self.left_barrier,
            *self.central_dot_gates,
            self.right_barrier,
        ]

    def query_neighbors(
        self,
        gate: "Gate",
    ) -> "Gates":
        """Queries the neighbors of a gate.

        Args:
            gate (Gate): The gate to query.

        Returns:
            Gates: The neighbors.
        """
        name = gate.name

        for gate_geometry in [
            *self.screening_gates,
            self.left_reservoir,
            self.right_reservoir,
            *self.all_dot_gates(),
        ]:
            if name != gate_geometry.value:
                continue
            if gate_geometry in self.screening_gates:
                return Gates(
                    [
                        self.left_barrier.left_neighbor,
                        *self.raw_gates,
                        self.right_barrier.right_neighbor,
                    ]
                )
            if gate_geometry == self.left_reservoir:
                return Gates(
                    [
                        self.left_reservoir.right_neighbor,
                        *self.screening_gates,
                    ]
                )
            if gate_geometry == self.right_reservoir:
                return Gates(
                    [
                        self.right_reservoir.left_neighbor,
                        *self.screening_gates,
                    ]
                )
            assert isinstance(
                gate_geometry,
                BarrierGateWithNeighbors | PlungerGateWithNeighbors,
            )
            return Gates(
                [
                    gate_geometry.left_neighbor,
                    gate_geometry.right_neighbor,
                    *self.screening_gates,
                ]
            )
        msg = f"Gate {name} not found in geometry."
        raise ValueError(msg)

File: cpp/src/physics/config/geometries/GateGeometryArray1D.cpp
#include "falcon_core/physics/config/geometries/GateGeometryArray1D.hpp"

#include <stdexcept>

#include "falcon_core/physics/config/geometries/DotGateWithNeighbors.hpp"
#include "falcon_core/physics/config/geometries/LeftReservoirWithImplantedOhmic.hpp"
namespace falcon_core {
namespace physics {
namespace config {
namespace geometries {
GateGeometryArray1D::GateGeometryArray1D(const GateGeometryArray1D& other) {
  std::unique_lock<std::shared_timed_mutex> lock_linear_array(_mu_linear_array,
                                                              std::defer_lock);
  std::unique_lock<std::shared_timed_mutex> lock_screening_gates(
      _mu_screening_gates, std::defer_lock);
  std::unique_lock<std::shared_timed_mutex> lock_raw_central_gates(
      _mu_raw_central_gates, std::defer_lock);
  std::unique_lock<std::shared_timed_mutex> lock_central_dot_gates(
      _mu_central_dot_gates, std::defer_lock);
  std::unique_lock<std::shared_timed_mutex> lock_gate_name_map(
      _mu_gate_name_map, std::defer_lock);
  std::lock(lock_linear_array,
            lock_screening_gates,
            lock_raw_central_gates,
            lock_central_dot_gates,
            lock_gate_name_map);
  if (!other._linear_array) {
    throw std::invalid_argument(
        "GateGeometryArray1D copy constructor: The lineararray cannot be "
        "null.");
  }
  _linear_array =
      std::make_shared<device_structures::Connections>(*other._linear_array);
  if (!other._screening_gates) {
    throw std::invalid_argument(
        "GateGeometryArray1D copy constructor: The screening_gates cannot be "
        "null.");
  }
  _screening_gates =
      std::make_shared<device_structures::Connections>(*other._screening_gates);
  _raw_central_gates = *std::make_shared<device_structures::Connections>(
      other._raw_central_gates);
  if (!other._central_dot_gates) {
    throw std::invalid_argument(
        "GateGeometryArray1D copy constructor: The central_dot_gates cannot be "
        "null.");
  }
  _central_dot_gates =
      std::make_shared<DotGatesWithNeighbors>(*other._central_dot_gates);
  _gate_name_map.clear();
  for (const auto& kv : other._gate_name_map) {
    _gate_name_map.emplace(kv.first, kv.second);
  }
}
GateGeometryArray1D GateGeometryArray1D::operator=(
    const GateGeometryArray1D& other) {
  if (this != &other) {
    std::unique_lock<std::shared_timed_mutex> lock_linear_array(
        _mu_linear_array, std::defer_lock);
    std::unique_lock<std::shared_timed_mutex> lock_screening_gates(
        _mu_screening_gates, std::defer_lock);
    std::unique_lock<std::shared_timed_mutex> lock_raw_central_gates(
        _mu_raw_central_gates, std::defer_lock);
    std::unique_lock<std::shared_timed_mutex> lock_central_dot_gates(
        _mu_central_dot_gates, std::defer_lock);
    std::unique_lock<std::shared_timed_mutex> lock_gate_name_map(
        _mu_gate_name_map, std::defer_lock);
    std::lock(lock_linear_array,
              lock_screening_gates,
              lock_raw_central_gates,
              lock_central_dot_gates,
              lock_gate_name_map);
    _linear_array =
        std::make_shared<device_structures::Connections>(*other._linear_array);
    _screening_gates = std::make_shared<device_structures::Connections>(
        *other._screening_gates);
    _raw_central_gates = *std::make_shared<device_structures::Connections>(
        other._raw_central_gates);
    _central_dot_gates =
        std::make_shared<DotGatesWithNeighbors>(*other._central_dot_gates);
    _gate_name_map.clear();
    for (const auto& kv : other._gate_name_map) {
      _gate_name_map.emplace(
          kv.first,
          std::make_shared<device_structures::Connection>(*kv.second));
    }
  }
  return *this;
}

GateGeometryArray1D::GateGeometryArray1D() = default;
GateGeometryArray1D::GateGeometryArray1D(
    device_structures::ConnectionsSP lineararray,
    device_structures::ConnectionsSP screening_gates)
    : _linear_array(lineararray),
      _screening_gates(screening_gates),
      _central_dot_gates(std::make_shared<DotGatesWithNeighbors>()) {
  if (!lineararray) {
    throw std::invalid_argument(
        "GateGeometryArray1D: The lineararray cannot be null.");
  }
  if (!screening_gates) {
    throw std::invalid_argument(
        "GateGeometryArray1D: The screening_gates cannot be null.");
  }
  if (!screening_gates->is_screening_gates()) {
    throw std::runtime_error(
        "GateGeometryArray1D: Expected the screening_gates to consist of only "
        "screening gates");
  }
  if (screening_gates->size() != 2) {
    throw std::invalid_argument(
        "GateGeometryArray1D: Expected two screening gates.");
  }
  if (lineararray->size() % 2 == 0) {
    throw std::invalid_argument(
        "GateGeometryArray1D: Expected an odd number of elements in the linear "
        "array. Got " +
        std::to_string(lineararray->size()) + " elements.");
  }
  if (!((*lineararray)[0]->is_ohmic()) ||
      !((*lineararray)[lineararray->size() - 1]->is_ohmic())) {
    throw std::invalid_argument(
        "GateGeometryArray1D: Expected Ohmic at the ends of the linear array.");
  }
  if (!((*lineararray)[1]->is_reservoir_gate()) ||
      !((*lineararray)[lineararray->size() - 2]->is_reservoir_gate())) {
    throw std::invalid_argument(
        "GateGeometryArray1D: Expected Reservoir Gates bounding the central "
        "dot gates.");
  }

  // Extract dot gates
  std::vector<device_structures::ConnectionSP> dot_gates;
  for (size_t i = 2; i < lineararray->size() - 2; ++i) {
    device_structures::ConnectionSP dot_gate = (*lineararray)[i];
    if (!(dot_gate->is_dot_gate())) {
      throw std::invalid_argument(
          "GateGeometryArray1D: Expected DotGates in the middle of the linear "
          "array.");
    }
    dot_gates.push_back(dot_gate);
  }
  _raw_central_gates = device_structures::Connections(dot_gates);

  if (!(dot_gates.front()->is_barrier_gate()) ||
      !(dot_gates.back()->is_barrier_gate())) {
    throw std::invalid_argument(
        "Expected Barrier Gates bounding the exterior reservoir gates.");
  }
  if (!(dot_gates[1]->is_plunger_gate()) ||
      !(dot_gates[dot_gates.size() - 2]->is_plunger_gate())) {
    throw std::invalid_argument(
        "Expected Plunger Gates at the correct positions.");
  }

  // Central dot gates logic (example, adjust as needed)
  for (size_t i = 1; i + 1 < dot_gates.size(); ++i) {
    auto left_neighbor  = dot_gates[i - 1];
    auto selected_gate  = dot_gates[i];
    auto right_neighbor = dot_gates[i + 1];
    append_central_gate(left_neighbor, selected_gate, right_neighbor);
  }

  // Build the gate name map
  auto s_gates = *this->screening_gates();
  for (const auto& screening_gate : s_gates) {
    _gate_name_map[screening_gate->name()] = screening_gate;
  }
  _gate_name_map[left_reservoir()->name()]  = left_reservoir();
  _gate_name_map[right_reservoir()->name()] = right_reservoir();
  auto all_dot                              = *all_dot_gates();
  for (const auto& dot_gate : all_dot) {
    _gate_name_map[dot_gate->name()] = dot_gate;
  }
}
GateGeometryArray1D::iterator GateGeometryArray1D::begin() {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_linear_array);
  return _linear_array->begin();
}
GateGeometryArray1D::iterator GateGeometryArray1D::end() {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_linear_array);
  return _linear_array->end();
}
GateGeometryArray1D::const_iterator GateGeometryArray1D::begin() const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_linear_array);
  return _linear_array->begin();
}
GateGeometryArray1D::const_iterator GateGeometryArray1D::end() const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_linear_array);
  return _linear_array->end();
}
const device_structures::ConnectionsSP& GateGeometryArray1D::lineararray()
    const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_linear_array);
  return _linear_array;
}
const device_structures::ConnectionsSP& GateGeometryArray1D::screening_gates()
    const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_screening_gates);
  if (!_screening_gates->is_screening_gates()) {
    throw std::runtime_error(
        "Expected the screening_gates to only be screening_gates");
  }

  return _screening_gates;
}
device_structures::ConnectionsSP GateGeometryArray1D::raw_central_gates()
    const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_raw_central_gates);
  return std::make_shared<device_structures::Connections>(_raw_central_gates);
}
DotGatesWithNeighborsSP GateGeometryArray1D::central_dot_gates() const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_central_dot_gates);
  return _central_dot_gates;
}
void GateGeometryArray1D::append_central_gate(
    const device_structures::ConnectionSP& left_neighbor,
    const device_structures::ConnectionSP& selected_gate,
    const device_structures::ConnectionSP& right_neighbor) {
  std::unique_lock<std::shared_timed_mutex> lock(_mu_central_dot_gates);
  if (!left_neighbor || !selected_gate || !right_neighbor) {
    throw std::invalid_argument(
        "GateGeometryArray1D: neighbors and selected_gate must not be null");
  }
  if (selected_gate->is_barrier_gate()) {
    if (!(left_neighbor->is_plunger_gate()) ||
        !(right_neighbor->is_plunger_gate())) {
      throw std::invalid_argument(
          "GateGeometryArray1D: Expected PlungerGate bounding selected "
          "BarrierGate.");
    }
    _central_dot_gates->push_back(
        DotGateWithNeighbors::BarrierGateWithNeighbors(
            selected_gate->name(), left_neighbor, right_neighbor));
  } else if (selected_gate->is_plunger_gate()) {
    if (!(left_neighbor->is_barrier_gate()) ||
        !(right_neighbor->is_barrier_gate())) {
      throw std::invalid_argument(
          "GateGeometryArray1D: Expected BarrierGate bounding selected "
          "PlungerGate.");
    }
    _central_dot_gates->push_back(
        DotGateWithNeighbors::PlungerGateWithNeighbors(
            selected_gate->name(), left_neighbor, right_neighbor));
  } else {
    throw std::invalid_argument(
        "Expected either a PlungerGate or BarrierGate.");
  }
}
DotGatesWithNeighborsSP GateGeometryArray1D::all_dot_gates() const {
  DotGatesWithNeighbors all_dot_gates;
  all_dot_gates.push_back(left_barrier());
  auto central_dot_gates = *this->central_dot_gates();
  for (const auto& gate : central_dot_gates) {
    all_dot_gates.push_back(gate);
  }
  all_dot_gates.push_back(right_barrier());
  return std::make_shared<DotGatesWithNeighbors>(all_dot_gates);
}
device_structures::ConnectionsSP GateGeometryArray1D::query_neighbors(
    const device_structures::ConnectionSP& gate) const {
  std::shared_lock<std::shared_timed_mutex> lock(_mu_gate_name_map);
  auto it = _gate_name_map.find(gate->name());
  if (it == _gate_name_map.end()) {
    throw std::invalid_argument("Gate " + gate->name() +
                                " not found in geometry.");
  }
  device_structures::ConnectionSP gate_geometry = it->second;

  device_structures::ConnectionsSP result =
      std::make_shared<device_structures::Connections>();

  // If gate_geometry is in screening_gates
  auto screening_gates = *this->screening_gates();
  for (const auto& sg : screening_gates) {
    if (sg->name() == gate_geometry->name()) {
      auto lbar      = left_barrier();
      auto lneighbor = lbar->left_neighbor();
      result->push_back(lneighbor);
      auto central_gates = raw_central_gates();
      for (const auto& g : *central_gates) {
        result->push_back(g);
      }
      auto rbar      = right_barrier();
      auto rneighbor = rbar->right_neighbor();
      result->push_back(rneighbor);
      return result;
    }
  }
  if (gate_geometry->name() == left_reservoir()->name()) {
    auto lres      = left_reservoir();
    auto rneighbor = lres->right_neighbor();
    result->push_back(rneighbor);
    for (const auto& sg : screening_gates) {
      result->push_back(sg);
    }
    return result;
  } else if (gate_geometry->name() == right_reservoir()->name()) {
    auto rres      = right_reservoir();
    auto lneighbor = rres->left_neighbor();
    result->push_back(lneighbor);
    for (const auto& sg : screening_gates) {
      result->push_back(sg);
    }
    return result;
  }
  auto gate_geom =
      std::dynamic_pointer_cast<DotGateWithNeighbors>(gate_geometry);
  result->push_back(gate_geom->left_neighbor());
  result->push_back(gate_geom->right_neighbor());
  for (const auto& sg : screening_gates) {
    result->push_back(sg);
  }
  return result;
}
LeftReservoirWithImplantedOhmicSP GateGeometryArray1D::left_reservoir() const {
  std::string                     name  = ((*lineararray())[1])->name();
  device_structures::ConnectionSP ohmic = (*lineararray())[0];
  device_structures::ConnectionSP barrier_gate =

      (*lineararray())[2];
  LeftReservoirWithImplantedOhmicSP left_reservoir =
      std::make_shared<LeftReservoirWithImplantedOhmic>(
          name, barrier_gate, ohmic);
  return left_reservoir;
}
RightReservoirWithImplantedOhmicSP GateGeometryArray1D::right_reservoir()
    const {
  std::string name = (*lineararray())[lineararray()->size() - 2] -> name();
  device_structures::ConnectionSP ohmic =
      (*lineararray())[lineararray()->size() - 1];
  device_structures::ConnectionSP barrier_gate =
      (*lineararray())[lineararray()->size() - 3];
  return std::make_shared<RightReservoirWithImplantedOhmic>(
      name, barrier_gate, ohmic);
}
DotGateWithNeighborsSP GateGeometryArray1D::left_barrier() const {
  std::string                     name = ((*lineararray())[2])->name();
  device_structures::ConnectionSP reservoir_gate = (*lineararray())[1];
  device_structures::ConnectionSP plunger_gate   = (*lineararray())[3];
  return DotGateWithNeighbors::BarrierGateWithNeighbors(
      name, reservoir_gate, plunger_gate);
}
DotGateWithNeighborsSP GateGeometryArray1D::right_barrier() const {
  std::string name = ((*lineararray())[lineararray()->size() - 3])->name();
  device_structures::ConnectionSP reservoir_gate =
      (*lineararray())[lineararray()->size() - 2];
  device_structures::ConnectionSP plunger_gate =
      (*lineararray())[lineararray()->size() - 4];
  return DotGateWithNeighbors::BarrierGateWithNeighbors(
      name, plunger_gate, reservoir_gate);
}
device_structures::ConnectionsSP GateGeometryArray1D::ohmics() const {
  device_structures::ConnectionSP left_ohmic = (*lineararray())[0];
  device_structures::ConnectionSP right_ohmic =
      (*lineararray())[lineararray()->size() - 1];
  std::vector<device_structures::ConnectionSP> tmp({left_ohmic, right_ohmic});
  return std::make_shared<device_structures::Connections>(tmp);
}
bool GateGeometryArray1D::operator==(const GateGeometryArray1D& other) const {
  return (*this->lineararray() == *other.lineararray()) &&
         (*this->screening_gates() == *other.screening_gates());
}
bool GateGeometryArray1D::operator!=(const GateGeometryArray1D& other) const {
  return !(*this == other);
}

}  // namespace geometries
}  // namespace config
}  // namespace physics
}  // namespace falcon_core
CEREAL_REGISTER_TYPE(
    falcon_core::physics::config::geometries::GateGeometryArray1D)
CEREAL_REGISTER_POLYMORPHIC_RELATION(
    falcon_core::generic::Song,
    falcon_core::physics::config::geometries::GateGeometryArray1D)
